### PR TITLE
[C#] Add complete pending to device API

### PR DESCRIPTION
**Pull request by badrishc**
So that we can ensure no outstanding IO when disposing allocator.

### CODE REVIEW

File: cs/src/core/Allocator/GenericAllocator.cs [hunk 1]
```diff
@@ -191,6 +191,13 @@ internal override bool TryComplete()
         /// </summary>
         public override void Dispose()
         {
+            // Dispose base allocator
+            base.Dispose();
```
Comment (TedHartMS) [inline cs/src/core/Allocator/GenericAllocator.cs:195]: It's generally better to destruct the base after derived. Did this need to move here?

**Review by TedHartMS — APPROVED** [review]

File: cs/src/core/Allocator/GenericAllocator.cs [hunk 2]
```diff
@@ -191,6 +191,13 @@ internal override bool TryComplete()
...
+
+            // Ensure pending IO on object log device is completed
+            objectLogDevice.CompletePending();
```
Comment (tli2) [inline cs/src/core/Allocator/GenericAllocator.cs:198]: This is probably a left over issue from before, but why does Dispose() here not call into dispose of objectLogDevice? It seems wrong to call CompletePending() without disposing, as that does not reliable stop future calls.

**Review by tli2 — APPROVED** [review]
Otherwise LGTM

File: cs/src/core/Allocator/GenericAllocator.cs (hunk 2 above)
Comment (badrishc) [reply cs/src/core/Allocator/GenericAllocator.cs:198]: We cannot dispose the device as we did not create the device, the user did. So they will have to dispose it.

**Comment by badrishc** [comment]
Closing - instead we will update the device to catch and ignore exceptions in the uncommon race condition of IO after allocator dispose.

**Pull request closed without merging**